src/DataStructures/Tensor/Expressions/TensorExpression.hpp

```diff
@@ -31,23 +62,49 @@
  * `ti_a` and `ti_b` are place holders for objects of type `TensorIndex<0>` and
  * `TensorIndex<1>` respectively.
  */
-template <std::size_t I>
+template <std::size_t I, Requires<(I < max_sentinel)> = nullptr>
 struct TensorIndex {
   using value_type = std::size_t;
   using type = TensorIndex<I>;
   static constexpr value_type value = I;
+  static constexpr UpLo valence =
+      ((I < upper_sentinel) or
+       (I >= spatial_sentinel and I < upper_spatial_sentinel))
+          ? UpLo::Lo
+          : UpLo::Up;
+  static constexpr bool is_spacetime = I < spatial_sentinel;
 };
 
-/// \ingroup TensorExpressionsGroup
-/// Given an `TensorIndex<size_t>` return `TensorIndex<size_t + 1>`
-/// \tparam T the args to increment
-template <typename T>
-using next_tensor_index = TensorIndex<T::value + 1>;
-
-/// \ingroup TensorExpressionsGroup
-/// Metafunction to return the sum of two TensorIndex's
-template <typename A, typename B>
-using plus_tensor_index = TensorIndex<A::value + B::value>;
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Returns the TensorIndex value of with opposite valence.
+ *
+ * \details The input value represents a TensorIndex value, which encodes
+ * both the valence of the index and whether the index is spacetime or
+ * spatial. This function returns the value that corresponds to the encoding of
+ * the TensorIndex with the same index type, but opposite valence.
+ *
+ * For example, 0 is the TensorIndex value for `ti_a_t`. If `i == 0`,
+ * then 500 will be returned, which is the TensorIndex value for `ti_A_t`.
+ * If `i == 500` (representing `ti_A_t`), then 0 (representing `ti_a_t`) is
+ * returned.
+ *
+ * @param i a TensorIndex value that represents a generic index
+ * @return the TensorIndex value that encodes the generic index with the
+ * opposite valence
+ */
+SPECTRE_ALWAYS_INLINE static constexpr size_t
+get_tensorindex_value_with_opposite_valence(const size_t i) noexcept {
+  assert(i < max_sentinel);
```

Comment:
Can you just add `  // NOLINT` to the end of this line to disable clang-tidy? As far as I can tell it's complaining about code in the standard library and we can't exactly change that... Feel free to squash that in immediately

Reply:
So to explain a bit on why the code is actually "okay" and an alternative that is more visibly okay we need to know some rules... C++ always constexpr evaluation as long as the _call path_ is constexpr. That is, you can have

```cpp
constexpr size_t blah(const size_t i) {
  if (i >= 10) {
    std::cout << "Yoohoo! Big summer blowout!\n";
  }
  return i;
}
```
and if `i < 10` this can be evaluated constexpr, otherwise you get the following "fun":
```shell
test.cpp:67:28:   in ‘constexpr’ expansion of ‘blah(10)’
test.cpp:59:18: error: call to non-‘constexpr’ function ‘std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const char*) [with _Traits = std::char_traits<char>]’
   59 |     std::cout << "Yoohoo! Big summer blowout!\n";
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/10/iostream:39,
                 from test.cpp:6:
/usr/include/c++/10/ostream:606:5: note: ‘std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const char*) [with _Traits = std::char_traits<char>]’ declared here
  606 |     operator<<(basic_ostream<char, _Traits>& __out, const char* __s)
      |     ^~~~~~~~
```
So basically, inside `assert` we have code like:
```cpp
if (not (ARGUMENT)) {
  printf("Blah blah blah");
  std::abort();
}
```
The non-constexpr code is not evaluated _unless_ `ARGUMENT` is `false`, which causes the `assert` to fail. This, of course, relies on the implementation-defined internals of `assert` (though I'm not sure how exactly you could have an implementation that's sufficiently weird that it wouldn't behave this way). A safer way to do the error handling would be:
```cpp
constexpr size_t blah(const size_t i) {
  if (i >= 10) {
    throw std::domain_error{"Value i must be less than 10" + std::to_string(i)};
  }
  return i;
}
```
The disadvantage compared to the runtime error always being you don't actually get to know the value of `i` but I'm unaware of a way to fix that.

Reply:
Thanks for the example @nilsdeppe!

@macedo22 I think I agree with Nils's suggestion on NOLINT'ing the assert. I don't think it makes sense to increase the complexity of the code here by adding templates, for instances.

Reply:
Thanks for the example, Nils! I also just played around with this locally for a few different scenarios. Makes sense. I've gone ahead and just suppressed the linting warning